catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java

```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
-		Project currentProject = ProjectManager.getInstance().getCurrentProject();
-
-		ProjectAndSceneScreenshotLoader screenshotLoader = new ProjectAndSceneScreenshotLoader(this);
-		screenshotLoader.loadAndShowScreenshot(currentProject.getName(),
-				currentProject.getDefaultScene().getName(),
-				false,
-				(ImageView) findViewById(R.id.project_image_view));
-
-		((TextView) findViewById(R.id.project_size_view)).setText(FileMetaDataExtractor.getSizeAsString(
-				new File(PathBuilder.buildProjectPath(currentProject.getName())), this));
+	private void onNextButtonClick() {
+		setNextButtonEnabled(false);
 
-		nameInputLayout = findViewById(R.id.input_project_name);
-		descriptionInputLayout = findViewById(R.id.input_project_description);
+		if (Utils.isDefaultProject(project, this)) {
+			nameInputLayout.setError(getString(R.string.error_upload_default_project));
+			nameInputLayout.getEditText().removeTextChangedListener(nameInputTextWatcher);
+			nameInputLayout.setEnabled(false);
+			descriptionInputLayout.setEnabled(false);
+			return;
+		}
 
-		nameInputLayout.getEditText().setText(currentProject.getName());
-		descriptionInputLayout.getEditText().setText(currentProject.getDescription());
+		setShowProgressBar(true);
+		showSelectTagsDialog();
+	}
 
-		nameInputLayout.getEditText().addTextChangedListener(textChangedListener);
-		invalidateOptionsMenu();
+	private void showSelectTagsDialog() {
+		final List<String> checkedTags = new ArrayList<>();
+		final String[] availableTags = tags.toArray(new String[tags.size()]);
+
+		new AlertDialog.Builder(this)
+				.setTitle(R.string.upload_tag_dialog_title)
+				.setMultiChoiceItems(availableTags, null, new DialogInterface.OnMultiChoiceClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int indexSelected, boolean isChecked) {
+						if (isChecked) {
+							if (checkedTags.size() >= MAX_NUMBER_OF_TAGS_CHECKED) {
+								((AlertDialog) dialog).getListView().setItemChecked(indexSelected, false);
+							} else {
+								checkedTags.add(availableTags[indexSelected]);
+							}
+						} else {
+							checkedTags.remove(availableTags[indexSelected]);
+						}
+					}
+				})
+				.setPositiveButton(getText(R.string.next), new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						project.setTags(checkedTags);
+						showProgressDialogAndUploadProject();
+					}
+				})
+				.setNegativeButton(getText(R.string.cancel), new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						Utils.invalidateLoginTokenIfUserRestricted(ProjectUploadActivity.this);
+						setShowProgressBar(false);
+						setNextButtonEnabled(true);
+					}
+				})
+				.setCancelable(false)
+				.show();
 	}
 
-	private void onNextButtonClick() {
-		String name = nameInputLayout.getEditText().getText().toString().trim();
+	private void showProgressDialogAndUploadProject() {
```

Comment:
this function is quite long

Reply:
yes, so what do want me to do about it? IMO it is clear what the snippet does (i.e. it get the `String name` from `nameInputLayout`.